### ISSUE TITLE
Fix DelegateClass

### ISSUE THREAD

**Issue by adambeynon** [opal/opal#605]
DelegateClass in stdlib/delegate.rb

**Comment by elia** [comment]
Current implementation:

```rb
def DelegateClass(superklass)
  SimpleDelegator
end
```

MRI 2.2 source:

```rb
def DelegateClass(superclass)
  klass = Class.new(Delegator)
  methods = superclass.instance_methods
  methods -= ::Delegator.public_api
  methods -= [:to_s,:inspect,:=~,:!~,:===]
  klass.module_eval do
    def __getobj__  # :nodoc:
      unless defined?(@delegate_dc_obj)
        return yield if block_given?
        __raise__ ::ArgumentError, "not delegated"
      end
      @delegate_dc_obj
    end
    def __setobj__(obj)  # :nodoc:
      __raise__ ::ArgumentError, "cannot delegate to self" if self.equal?(obj)
      @delegate_dc_obj = obj
    end
    methods.each do |method|
      define_method(method, Delegator.delegating_block(method))
    end
  end
  klass.define_singleton_method :public_instance_methods do |all=true|
    super(all) - superclass.protected_instance_methods
  end
  klass.define_singleton_method :protected_instance_methods do |all=true|
    super(all) | superclass.protected_instance_methods
  end
  return klass
end
```

**Comment by hmdne** [comment]
#2274 maybe?

**Comment by sarhm** [comment]
> DelegateClass in stdlib/delegate.rb
>

**Comment by hmdne** [comment]
This has been fixed with #2446
